# Life Expectancy Prediction Project

I will use the **Life Expectancy dataset**, collected by the WHO and the United Nations, assembled and available at: https://www.kaggle.com/datasets/kumarajarshi/life-expectancy-who. The columns of the dataset are:

* **Life expectancy:** measured in years (target variable)
* **Country**
* **Year**
* **Status:** developing or developed country
* **Population:** population of the country
* **Adult mortality:** probability of dying between 15 and 60 years per 1000 population, for both sexes
* **Infant deaths:** number of infant deaths per 1000 population
* **Under-five deaths:** number of under-five deaths per 1000 population
* **GDP:** Gross Domestic Product per capita (in USD)
* **Percentage expediture:** expenditure on health as a percentage of Gross Domestic Product per capita (%)
* **Total expediture:** general government expenditure on health as a percentage of total government expenditure (%)
* **Income composition:** Human Development Index in terms of income composition of resources (index ranging from 0 to 1)
* **Hepatitis B:** hepatitis B (HepB) immunization coverage among 1-year-olds (%)
* **Polio:** polio (Pol3) immunization coverage among 1-year-olds (%)
* **Diphteria:** Diphtheria tetanus toxoid and pertussis (DTP3) immunization coverage among 1-year-olds (%)
* **Measles:** number of reported cases per 1000 population
* **HIV/AIDS:** deaths per 1000 live births HIV/AIDS (0-4 years)
* **Thinness 5-9:** prevalence of thinness among children and adolescents for age 5 to 9 (%)
* **Thinness 10-19:** prevalence of thinness among children and adolescents for age 10 to 19 (%)
* **BMI:** average Body Mass Index of entire population
* **Alcohol:** recorded per capita (15+) consumption (in litres of pure alcohol)
* **Schooling:** number of years of schooling

In [1]:
import pandas as pd
pd.set_option('display.max_rows', 500)
import numpy as np
np.set_printoptions(legacy='1.25')
import plotly.io as pio
import plotly.express as px
pio.templates.default = "simple_white"
from sklearn.model_selection import train_test_split

In [2]:
df = pd.read_csv("life_expectancy_data.csv")
df

Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01               71.279624         65.0      1154   
1                64     0.01               73.523582         62.0       492   
2                66     0.01               73.219243         64.0       430   
3                69     0.01               78.184215         67.0      2787   
4                71     0.01                7.097109         68.0      3013   
...             ...      ...                     ...          ...       ...   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
0     ...    6.0               8.16         65.0        0.1  584.259210   
1     ...   58.0               8.18         62.0        0.1  612.696514   
2     ...   62.0               8.13         64.0        0.1  631.744976   
3     ...   67.0               8.52         67.0        0.1  669.959000   
4     ...   68.0               7.87         68.0        0.1   63.537231   
...   ...    ...                ...          ...        ...         ...   
2933  ...   67.0               7.13         65.0       33.6  454.366654   
2934  ...    7.0               6.52         68.0       36.7  453.351155   
2935  ...   73.0               6.53         71.0       39.8   57.348340   
2936  ...   76.0               6.16         75.0       42.1  548.587312   
2937  ...   78.0               7.10         78.0       43.5  547.358878   

      Population   thinness  1-19 years   thinness 5-9 years  \
0     33736494.0                   17.2                 17.3   
1       327582.0                   17.5                 17.5   
2     31731688.0                   17.7                 17.7   
3      3696958.0                   17.9                 18.0   
4      2978599.0                   18.2                 18.2   
...          ...                    ...                  ...   
2933  12777511.0                    9.4                  9.4   
2934  12633897.0                    9.8                  9.9   
2935    125525.0                    1.2                  1.3   
2936  12366165.0                    1.6                  1.7   
2937  12222251.0                   11.0                 11.2   

      Income composition of resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                               0.470        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427      

# Exploratory Data Analysis

In [3]:
# Formatting column names:
df.columns = df.columns.str.lower().str.strip()
df.columns = df.columns.str.replace(' ', '_').str.replace('__', '_').str.replace('/', '_')
df = df.rename(columns={'thinness_1-19_years': 'thinness_10-19_years'})

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   country                          2938 non-null   object 
 1   year                             2938 non-null   int64  
 2   status                           2938 non-null   object 
 3   life_expectancy                  2928 non-null   float64
 4   adult_mortality                  2928 non-null   float64
 5   infant_deaths                    2938 non-null   int64  
 6   alcohol                          2744 non-null   float64
 7   percentage_expenditure           2938 non-null   float64
 8   hepatitis_b                      2385 non-null   float64
 9   measles                          2938 non-null   int64  
 10  bmi                              2904 non-null   float64
 11  under-five_deaths                2938 non-null   int64  
 12  polio               

In [5]:
# Check if there are duplicated data:
df.duplicated().any()

False

In [6]:
cols_numerical = df.select_dtypes(include=['int64', 'float64']).columns
cols_categorical = df.select_dtypes(include=['object']).columns

In [7]:
# Value counts of categorical data:
for col in cols_categorical:
    print('Number of unique values of', col, ':', df[col].nunique())
    value_counts = df[col].value_counts()
    print(value_counts.value_counts())
    print()

Number of unique values of country : 193
count
16    183
1      10
Name: count, dtype: int64

Number of unique values of status : 2
count
2426    1
512     1
Name: count, dtype: int64



In [8]:
# Summary stats of numerical data:
df.describe().T.apply(lambda s: s.apply('{0:.2f}'.format))

count         mean          std      min  \
year                             2938.00      2007.52         4.61  2000.00   
life_expectancy                  2928.00        69.22         9.52    36.30   
adult_mortality                  2928.00       164.80       124.29     1.00   
infant_deaths                    2938.00        30.30       117.93     0.00   
alcohol                          2744.00         4.60         4.05     0.01   
percentage_expenditure           2938.00       738.25      1987.91     0.00   
hepatitis_b                      2385.00        80.94        25.07     1.00   
measles                          2938.00      2419.59     11467.27     0.00   
bmi                              2904.00        38.32        20.04     1.00   
under-five_deaths                2938.00        42.04       160.45     0.00   
polio                            2919.00        82.55        23.43     3.00   
total_expenditure                2712.00         5.94         2.50     0.37   
diphtheria                       2919.00        82.32        23.72     2.00   
hiv_aids                         2938.00         1.74         5.08     0.10   
gdp                              2490.00      7483.16     14270.17     1.68   
population                       2286.00  12753375.12  61012096.51    34.00   
thinness_10-19_years             2904.00         4.84         4.42     0.10   
thinness_5-9_years               2904.00         4.87         4.51     0.10   
income_composition_of_resources  2771.00         0.63         0.21     0.00   
schooling                        2775.00        11.99         3.36     0.00   

                                       25%         50%         75%  \
year                               2004.00     2008.00     2012.00   
life_expectancy                      63.10       72.10       75.70   
adult_mortality                      74.00      144.00      228.00   
infant_deaths                         0.00        3.00       22.00   
alcohol                               0.88        3.75        7.70   
percentage_expenditure                4.69       64.91      441.53   
hepatitis_b                          77.00       92.00       97.00   
measles                               0.00       17.00      360.25   
bmi                                  19.30       43.50       56.20   
under-five_deaths                     0.00        4.00       28.00   
polio                                78.00       93.00       97.00   
total_expenditure                     4.26        5.75        7.49   
diphtheria                           78.00       93.00       97.00   
hiv_aids                              0.10        0.10        0.80   
gdp                                 463.94     1766.95     5910.81   
population                       195793.25  1386542.00  7420359.00   
thinness_10-19_years                  1.60        3.30        7.20   
thinness_5-9_years                    1.50        3.30        7.20   
income_composition_of_resources       0.49        0.68        0.78   
schooling                            10.10       12.30       14.30   

                                           max  
year                                   2015.00  
life_expectancy                          89.00  
adult_mortality                         723.00  
infant_deaths                          1800.00  
alcohol                                  17.87  
percentage_expenditure                19479.91  
hepatitis_b                              99.00  
measles                              212183.00  
bmi                                      87.30  
under-five_deaths                      2500.00  
polio                                    99.00  
total_expenditure                        17.60  
diphtheria                               99.00  
hiv_aids                                 50.60  
gdp                                  119172.74  
population                       1293859294.00  
thinness_10-19_years                     27.70  
thinness_5-9_years    

In [ ]:
# Histograms for all columns:
for col in df.columns:
    fig = px.histogram(df, x=col)
    fig.show()

When inspecting the summary stats and histograms, several columns had odd or peculiar data:

* Although the majority of countries have data for the 16 years (2000-2015), 10 countries only have data for year 2013.
* `infant_deaths`, `under-five_deaths` and `measles` contains values >1000. These may represent total cases, instead of cases per 1000 population.
* `population` showed values as low as 34 and as high as 1.3 billion.
* `percentage_expediture` does not cointain values between 0-1 or 0-100 (percentages), thus it may not be corrected by GDP. However, `total_expenditure` seems to have correct values and gives similar information as `percentage_expediture`.
* `income_composition_of_resources` contains many 0 values, which could represent missing values, as there is a gap of values >0 and <0.25.
* `bmi` shows values unrealistically low (<18.5) or high (>39.9), outside of the natural ranges of BMI.
* Also, the distribution of prevalence of thinness in children between 5-9 and 10-19 was very similar.

We'll have a deeper look into some of these columns:

* Fill missing values after filtering rows
* Data transformation
* Bivariate/Multivariate analysis

## Data cleaning